### PR TITLE
Generalize TPeerId into TConnInfo

### PR DESCRIPTION
**Pull request by tomaka**
This is the first part of #1030 

At the moment, the structs in the `nodes` module have a `TPeerId` parameter that notably requires `Eq + Hash`.

This PR turns `TPeerId` into `TConnInfo`, and introduces a new trait that `TConnInfo` must implement:

```rust
/// Information about a connection.
pub trait ConnectionInfo {
    /// Identity of the node we are connected to.
    type PeerId: Eq + Hash;

    /// Returns the identity of the node we are connected to on this connection.
    fn peer_id(&self) -> &Self::PeerId;
}

impl ConnectionInfo for PeerId {
    type PeerId = PeerId;

    fn peer_id(&self) -> &PeerId {
        self
    }
}
```

By default, a `TConnInfo` is a `PeerId`. But after this PR, it can consist of anything that contains a `PeerId`.

### CODE REVIEW

File: core/src/nodes/collection.rs [hunk 1]
```diff
@@ -143,65 +143,85 @@ where TOutEvent: fmt::Debug,
...
-    #[inline]
-    pub fn peer_id(&self) -> &TPeerId {
-        &self.peer_id
+    /// Returns the information of the connection.
+    pub fn connection_info(&self) -> &TConnInfo {
+        self.conn_info.as_ref().expect("conn_info is always Some when the object is alive; QED")
+    }
+
+    /// Returns the identity of the node we connected to.
+    pub fn peer_id(&self) -> &TPeerId
```
Comment (twittner) [inline core/src/nodes/collection.rs]: How about removing `peer_id()`?

Comment (tomaka) [reply core/src/nodes/collection.rs]: I had it for convenience, as a "why not have it" item, but I'll remove it.

Comment (tomaka) [reply core/src/nodes/collection.rs]: It is there for convenience and is used a lot in `raw_swarm`. Can I keep it? :pray:

Comment (twittner) [reply core/src/nodes/collection.rs]: Of course you can keep it :smile:. It was just a thought.

**Pull request merged**